File: 487/max-consecutive-ones-ii.py
# 這題的解法，等價於：能接受 one 最多有一個 的 two-pointer 題目
class Solution:
    def findMaxConsecutiveOnes(self, nums: List[int]) -> int:
        N = len(nums)
        left, right = 0, 0 # 右不包含
        zero, one, ans = 0, 0, 0
        while right<N: # 右邊還沒撞到邊界
            if nums[right]==1: # 是1，太好了
                right += 1
                one += 1
            elif nums[right]==0 and zero==0: # 是0也是可以啦，但只有1個quota
                right += 1
                zero += 1
            else: # 這是 zero超標的狀況
                right += 1
                zero += 1 # 很遺憾，現在 zero 超標了
                while zero > 1: # 只要超標，就持續吐出 nums[left]
                    if nums[left]==0:
                        zero -= 1 # 吐出 zero
                    else:
                        one -= 1 # 吐出 one
                    left += 1 # 吐出後，改變 left 的位置
            # print("left:", left, " right:", right, " zero:", zero, " one:", one)
            if one + zero > ans:
                ans = one + zero        
        return ans

File: 487/max-consecutive-ones-ii.cpp
// LeetCode 487. Max Consecutive Ones II
// nums 裡，若能把「1個0變1」，最多能「連續幾個1」
// 可用「毛毛蟲」的方法，容忍「最多1個0」來爬行即可
class Solution {
public:
    int findMaxConsecutiveOnes(vector<int>& nums) {
        int zero = 0, left = 0, ans = 0;
        for(int right = 0; right < nums.size(); right++) { // 右邊的頭往右爬
            if(nums[right]==0) { // 遇到0，就要處理 
                zero++; // 多了1個0
                while(zero>1) { // 0太多，就一直縮左邊的尾巴
                    if(nums[left]==0) zero--; // 太好了，吐出1個0
                    left++; // 左邊的尾巴往右縮
                }
            } 
            ans = max(ans, right-left+1); // 最長的長度在這裡
        }
        return ans;
    }
};
